runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -965,18 +966,24 @@ private void scheduleWorkItem(
         WindmillTimeUtils.windmillToHarnessWatermark(workItem.getOutputDataWatermark());
     Preconditions.checkState(
         outputDataWatermark == null || !outputDataWatermark.isAfter(inputDataWatermark));
+    WorkProcessingContext workProcessingContext =
+        WorkProcessingContext.builder(
+                computationState.getComputationId(),
+                (computationId, request) ->
+                    Optional.ofNullable(
+                        metricTrackingWindmillServer.getStateData(computationId, request)))
+            .setWorkCommitter(workCommitter::commit)
+            .setOutputDataWatermark(outputDataWatermark)
+            .setInputDataWatermark(inputDataWatermark)
+            .setSynchronizedProcessingTime(synchronizedProcessingTime)
+            .setWorkItem(workItem)
+            .build();
     Work scheduledWork =
```

Comment:
We now have WorkItem, WorkProcessingContext, and Work classes.
Does the ProcessingContext benefit from being separate from Work?
If they are valuable distinctions can we think of better names to explain them?
Or perhaps we can have different interfaces and a single class implementing them?

It's odd that process() is a Work member instead of WorkProcessingContext which has commit.

Reply:
updated
added `Watermarks` and `ProcessingContext` as subclasses of `Work` and all interactions for work processing goes through `Work.java`